### ISSUE TITLE
[RuTracker] Cleaning titles for Sonarr

### ISSUE THREAD

**Issue by bormanua** [Jackett/Jackett#1652]
In order for Sonarr to properly parse titles, Jackett will need to clean that up as it's not standard to have that there.

Example: 17-8-11 15:49:50.8|Debug|Parser|Parsing string ' / The Last Ship [S01-03] (2014-2016) WEB-DLRip Generalfilm | | LostFilm'

[Sonarr closed issue](https://github.com/Sonarr/Sonarr/issues/2117)

Jackett Version 0.7.1662.0
Mono Version 4.6.2

-------------------------------

**Comment by kaso17** [comment]
The real title is "Последний корабль / The Last Ship [S01-03] (2014-2016) WEB-DLRip от Generalfilm | КПК | LostFilm", sonarr strips the non ASCII characters.
As seventor apparently doesn't use scene release names sonarr will have a hard time parsing it.
is the naming scheme of seventor always "russian name / english name"?

**Comment by bormanua** [comment]
https://rutracker.org/forum/viewtopic.php?t=479259#9

Here is the naming rules (Google translate):

The order of the title of the topic in the Foreign TV series
Although the template itself makes the title of the theme, but still many or themselves re-themes or make out without a template so the title of the topic does not comply with the rules!

The name of the series in Russian Language / The title of the series in the Original Language / Season / With which and for what series (from how many) (Director's name) [Year of release, Country, Genre, Quality] Additional information on the release *

* Additional information on the release is:
Type / s
Presence of subtitles
Author / s translation / s (studio, title WG or nickname of the release)

Optionally, you can also specify other information that does not conflict with the rules of the section (important differences in the release that the author wants to emphasize)

For example:

Lost / Call of Blood / Fairy / Lost Girl / Season: 2 / Series: 1-6 (22) (Erik Canuel, Robert Lieberman, Paul Fox) [2011, USA, Fantasy, WEB-DLRip] MVO (Smart's Studios) + Rus Subs

_Sent from my Motorola Moto G (5) Plus using [FastHub](https://play.google.com/store/apps/details?id=com.fastaccess.github)_

**Comment by JigSawFr** [comment]
In my mind, we can clean up these titles in indexer definition for rutracker
What do you think of this @kaso17  ?

**Comment by kaso17** [comment]
I've no problem is someone adds a corresponding option to rewrite the titles.

**Comment by bormanua** [comment]
Can someone add this rewriting option? Please.

**Comment by bormanua** [comment]
Can I sponsor this rewrite option?

**Comment by kaso17** [comment]
@sv01a made some improvements as part of #2151
maybe that solves your problems.

**Comment by enchained** [comment]
Could someone possibly adjust the way that titles of the TV shows are shown to Sonarr?
In Rutracker the name of the torrent ends with "voiceover number and type" "voiceover release groups" + "Original" + "subs presense, language, authors".
Each of those parts can be optional, and presence of those parts for sonarr is crucial if for example some prefer specific realese group, specific subs, or just any subs presence.
Right now the way it parses is a bag of mixed goods. I'd like to help any way I can in testing, examples and usecases, but I'm afraid I'm no good at universal regexps, and this stuff can come pretty varied due to human factor.

So here we go:
Right after the quality i.e. ...`BDRip 1080p]` goes the translation info.
### 1. Russian voiceover part examples (MVO, DVO etc are types by number of actor voices/quality of VO studio):
`DVO (Gears Media)`
`MVO (Jaskier, NewStudio)`
`3xMVO (Jaskier, LostFilm, NewStudio) ` 
`MVO (Jaskier, LostFilm, NewStudio)` _yeah, sometimes they write 3x, 2x, 5x etc, sometimes they don't_
or it can be completely missing (like if it's translated with subs. TV without ANY translation to russian is not allowed on rutracker, so it has either voiceover translation, or subs translation, or both, but never neither - except for russian-originated content sections).

### 2. Original track presence:
`Original`
This one can be missing too. So there are such cases:
`<voiceover> + Original + <subs>`
`<voiceover> + <subs>` or just  `<subs>` - _pretty sure original is just ommited there, cause of subs make no sence without original track, but it's pretty rare form nowdays_
`<voiceover> + Original`
`Original + <subs>`
It can't just be "Original" tho, cause russian shows that are in separate TV subcategories, don't have that section of voiceover/original/subs at all. It just ends on the video quality.

### 3. Subtitles:
They are often missing, in older releases especially, but if they are mentioned it can vary:
Subs (Rus, Ukr, Eng)
Subs (2xRus, Eng)
Rus Subs (alesandra) + Eng Subs - example with subs release group. but often they don't have one
Eng Subs
Rus Subs
Rus subs (OCRus)
Rus (N-Team) & Eng Subs
Subs (Rus - Goblin, Eng)
rus sub

### 4. All together:
https://i.imgur.com/BC9w4Fh.png
I even found an outlier in recent releases, which goes like that:
`Тайный круг / The Secret Circle / Сезон: 1 / Серии: 1-22 из 22 (Лиз Фридлендер / Liz Friedlander) [2011, США, Канада, драма, фэнтези, WEB-DL 720p] (LostFilm MVO | Кубик в Куб DVO | Original + Sub)`
There's a lot of mess with older releases too, cause it was before the strict template rules and the semi-auto-generation of the thread title I guess.

All in all, it looks like it would be really hard to parse properly, and for now jackett confusing or omitting those parts for me:
https://i.imgur.com/bfOeMDp.png
**Could it be at least returned JUST THE WAY AS IT IS after the square bracket of the video quality?** So one could apply keywords restrictions on sonarr knowing that release group info is consistent with what is shown on rutracker itself.

Also I think there is need for the uploader to be in the mix - it's in the column "Автор" from my rutracker screenshot (https://i.imgur.com/BC9w4Fh.png). I'll explain why: when there's ongoing series, the uploader updates his release thread with new torrent file every time he has a new episode to release, and renames the thread episodes count accordingly. Some uploaders use specific codecs, or put specific translations. Some can update it with propers. All in all, I'm just looking for a way to follow a specific rutracker thread and just redownload it as it updates it's torrent. I'm not sure if sonarr should/can handle this case (cause preferrably it shouldn't redownload the old episodes), and I'm not sure if jackett could handle it either. So the only solution I came up with to follow specific thread - using sonarr keywords to restrict to the uploader of the desired thread. Combined with quality profile and some release groups info keywords, that will lead to redownloading that specific thread torrent every time (in theory). If you have any better ideas for solution, please let me know. This behavior isn't just rutracker-specific. Many russian trackers do that for ongoings, like kinozal, noname club, and probably most others from Jackett list too.

**Comment by kaso17** [comment]
@sv01a can you have a look at the request from @enchained?

**Comment by enchained** [comment]
Tested Rutracker in Radarr, looks like it needs some cleaning for movies too:
`Ван Гог. С любовью, Винсент / Loving Vincent (Дорота Кобела / Dorota Kobiela, Хью Уэлшман / Hugh Welchman) [2017, Великобритания, Польша, драма, криминал, биография, мультфильм, HDRip] Dub`
turns into
`. , / Loving Vincent ( / Dorota Kobiela, / Hugh Welchman) [2017, , , , , , , HDRip] Dub`
with the error
`Failed to map movie, found title , / Loving Vincent ( / Dorota Kobiela, / Hugh Welchman) [, expected one of: Ван Гог. С любовью, Винсент, Loving Vincent`

So, for TorznabCatType.MoviesForeign only 1 foreign title should be extracted out of the construction `Russian Title 1 / Russian Title 2 / Original title` (probably the last one if multiple) similarly to how it's done now with TV Shows, then names in parenthesis should be ignored - (Director 1 / Director 2), then ignore everything non-English between the year an the quality (including commas). Then include everything about translation after the last square bracket.

Also I want to repeat the easiest TL;DR solution for TV Shows - include everything after the [Year and Quality of the Rip] unchanged about dubs and subs.

**Comment by i3laze** [comment]
Anyone to fix LostFilm title parsing?
Got 0 results on HomeLand, and The Magicians give me results with "Release rejected: Unknown Series".

**Comment by i3laze** [comment]
1. Does not return Russian title even if “Strip Russian letters” unchecked.
As stated earlier, Radarr cannot map title, cause of tracker also returns ( имя режиссёра / Director name). 
I guess everything in ( ) should be omitted by parser.

**Comment by edalex86** [comment]
Hey all.
I found out that rutracker series regex sometimes fails and I see unparsed title sometimes.
I've looked at regex and it searches brackets in additional info and fails if couldn't find them.
Sometimes those brackets are missing indeed and since it's not so strict I'm voting for @enchained 's suggestion to have them as-is.
regex will look like this I think:
`var regex = new Regex(".+\\/\\s([^а-яА-я\\/]+)\\s\\/.+Сезон\\s*[:]*\\s+(\\d+).+(?:Серии|Эпизод)+\\s*[:]*\\s+(\\d+-*\\d*).+,\\s+(.+)\\]\\s(.+)");`
Also torznab api returned wrong results with season param but I will investigate it later.

**Comment by edalex86** [comment]
@sv01a @kaso17 could you look at my series regex fixing title endings?
I can't compile and test myself since I have VS 2012. Honestly, I was thinking about recompiling jackett with .net 3.5/4.0 to let it run on xp :smile: 

And I don't confirm wrong result with season param. That was probably regex problem too.
Example: search "taken season 2" will find result like "Taken Season 01-02" which is ok but regex will convert its name to "Taken S01" which missled me. I didn't try to fix this case yet.

Also I thought about the situation described by @enchained about following specific thread and checking it for change and I've ended up with external code repeating search with the same keywords, searching thread url in results and checking if name or size has changed.

**Comment by kaso17** [comment]
@edalex86 updated via 08ad94a2f55f969e81de78427d358a005eb4ee8c, please test

**Comment by edalex86** [comment]
Thanks, @kaso17 !
Looks much better for me.  I hope others will share some thoughts too. Also maybe we could think about movies cleaning regex if it could applied/needed for Jackett.

**Comment by i3laze** [comment]
@kaso17 , could you please help me with this one?
https://github.com/Jackett/Jackett/issues/2631

**Comment by edalex86** [comment]
Movie regex (if needed) could look like this
**.+\/\s([^а-яА-я\/]+).*\(.+\).+?(\d{4}).+,\s+(.+)\]\s(.+)** and replacemet string is **$1 ($2) $3 $4** or **$1 $2 $3 $4**
I didn't escape symbols for c# . Need more testing, it's just a mockup

**Comment by i3laze** [comment]
Curious question:
RuTracker has lots of series releases for the seasons currently in progress.
People just update a torrent with additional episodes (and change it’s title accordingly).

Any way Jackett can match such scenario and return a positive episode search result to Sonarr?

**Comment by HippyRock** [comment]
Any updates on parsing only English name version for Radarr? 
Currently "Бездна / The Abyss" gets searched as "/ The Abyss" and that confuses mapping and adds extra not necessary "also know as" variation with extra "slash" at the beginning.

**Comment by imm21** [comment]
Same issue with kinozal.

**Comment by kaso17** [comment]
I don't use any rutracker myself so I don't see the problems.
Someone would have to make a selection of various release names/format examples and adjust/test the regexes accordingly.

There are various only services to test the regexes, e.g. http://regexstorm.net/tester

**Comment by imm21** [comment]
For RuTracker and NNM-club problem is solved. Need the [same](https://github.com/Jackett/Jackett/pull/2566) fixes for other (rus) trackers like kinozal and casstudio.

**Comment by garfield69** [comment]
closing as stale.

**Comment by ilike2burnthing** [comment]
May be covered by https://github.com/Jackett/Jackett/commit/14bcfcc0181f79677dc496678ba4e03d946b2c96
